x-pack/plugins/rule_registry/server/rule_data_plugin_service/rule_data_plugin_service.ts

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { ElasticsearchClient, Logger } from 'kibana/server';
+
+import { RuleDataClient } from '../rule_data_client';
+import { IndexNames } from './index_names';
+import { IndexOptions } from './index_options';
+import { ResourceNames } from './resource_names';
+import { ResourceInstaller, Resources } from './resource_installer';
+
+interface ConstructorOptions {
+  getClusterClient: () => Promise<ElasticsearchClient>;
+  logger: Logger;
+  isWriteEnabled: boolean;
+  index: string;
+}
+
+export class RuleDataPluginService {
+  private readonly resourceNames: ResourceNames;
+  private readonly resourceInstaller: ResourceInstaller;
+  private readonly installCommonResources: () => Promise<void>;
+
+  constructor(private readonly options: ConstructorOptions) {
+    this.resourceNames = new ResourceNames({ indexPrefixFromConfig: options.index });
+    this.resourceInstaller = new ResourceInstaller({
+      resourceNames: this.resourceNames,
+      getClusterClient: options.getClusterClient,
+      logger: options.logger,
+      isWriteEnabled: options.isWriteEnabled,
+    });
+
+    this.installCommonResources = this.resourceInstaller.memoizeInstallation(Resources.common, () =>
+      this.resourceInstaller.installResourcesSharedBetweenAllIndices()
+    );
```

Comment:
TL;DR: I can do that.

I think it's a good suggestion, although to me these 2 ways are quite similar. `resourceInstaller.memoizeInstallation()` guarantees that the resulting async function will run its body only once, and then return the same promise no matter how many times called, when and where.

> We could start the common resource installation automatically in the constructor

Yeah, we will need to do that if we want to replace memoized functions with simple Promises. In general, I don't think it's a good idea to start background processes (like a calculation or IO activity) from constructors. In this case, I think it should be ok since it's meant to be used only in once place, in the same plugin that we develop.

Reply:
> I'd recommend to keep the constructor free of side-effects. Separating construction and initialization is often important for unit-testing and subclassing.

I agree, I share the same feeling.

Reply:
Ok, @marshallmain @weltenwort I'll see if I can use pure Promises and keep the constructor side-effect free.